# Notebook 01 — Data Extraction & Initial Inspection
**Project:** Olist E-Commerce Performance Analysis  
**Purpose:** Load all raw CSV files, inspect structure, document shape, dtypes, and null counts  
**Output:** Summary inspection saved to docs — no data is modified here


In [1]:
import pandas as pd
import numpy as np

print("Libraries loaded successfully")


Libraries loaded successfully


In [2]:
orders       = pd.read_csv("olist_orders_dataset.csv")
order_items  = pd.read_csv("olist_order_items_dataset.csv")
customers    = pd.read_csv("olist_customers_dataset.csv")
payments     = pd.read_csv("olist_order_payments_dataset.csv")
reviews      = pd.read_csv("olist_order_reviews_dataset.csv")
products     = pd.read_csv("olist_products_dataset.csv")
sellers      = pd.read_csv("olist_sellers_dataset.csv")
geolocation  = pd.read_csv("olist_geolocation_dataset.csv")
category_translation = pd.read_csv("product_category_name_translation.csv")

print("All 9 tables loaded successfully")

All 9 tables loaded successfully


In [3]:
tables = {
    "orders": orders,
    "order_items": order_items,
    "customers": customers,
    "payments": payments,
    "reviews": reviews,
    "products": products,
    "sellers": sellers,
    "geolocation": geolocation,
    "category_translation": category_translation
}

print(f"{'Table':<25} {'Rows':>10} {'Columns':>10}")
print("-" * 47)
for name, df in tables.items():
    print(f"{name:<25} {df.shape[0]:>10,} {df.shape[1]:>10}")

Table                           Rows    Columns
-----------------------------------------------
orders                        99,441          8
order_items                  112,650          7
customers                     99,441          5
payments                     103,886          5
reviews                       99,224          7
products                      32,951          9
sellers                        3,095          4
geolocation                1,000,163          5
category_translation              71          2


In [4]:
print("=== SAMPLE DATA PREVIEW ===\n")
for name, df in tables.items():
    print(f"--- {name} ---")
    display(df.head(3))
    print()

=== SAMPLE DATA PREVIEW ===

--- orders ---


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00



--- order_items ---


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87



--- customers ---


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP



--- payments ---


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71



--- reviews ---


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24



--- products ---


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0



--- sellers ---


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ



--- geolocation ---


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP



--- category_translation ---


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto


In [5]:
print("=== DATAFRAME INFO ===\n")
for name, df in tables.items():
    print(f"--- {name} ---")
    df.info()
    print()

=== DATAFRAME INFO ===

--- orders ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB

--- order_items ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               -------------- 

In [6]:
print("=== STATISTICAL SUMMARY (Numerical Columns) ===\n")
for name, df in tables.items():
    numeric = df.select_dtypes(include='number')
    if not numeric.empty:
        print(f"--- {name} ---")
        display(numeric.describe().round(2))
        print()

=== STATISTICAL SUMMARY (Numerical Columns) ===

--- order_items ---


,order_item_id,price,freight_value
count,112650.00,112650.00,112650.00
mean,1.20,120.65,19.99
std,0.71,183.63,15.81
min,1.00,0.85,0.00
25%,1.00,39.90,13.08
50%,1.00,74.99,16.26
75%,1.00,134.90,21.15
max,21.00,6735.00,409.68



--- customers ---


,customer_zip_code_prefix
count,99441.00
mean,35137.47
std,29797.94
min,1003.00
25%,11347.00
50%,24416.00
75%,58900.00
max,99990.00



--- payments ---


,payment_sequential,payment_installments,payment_value
count,103886.00,103886.00,103886.00
mean,1.09,2.85,154.10
std,0.71,2.69,217.49
min,1.00,0.00,0.00
25%,1.00,1.00,56.79
50%,1.00,1.00,100.00
75%,1.00,4.00,171.84
max,29.00,24.00,13664.08



--- reviews ---


,review_score
count,99224.00
mean,4.09
std,1.35
min,1.00
25%,4.00
50%,5.00
75%,5.00
max,5.00



--- products ---


,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32341.00,32341.00,32341.00,32949.00,32949.00,32949.00,32949.00
mean,48.48,771.50,2.19,2276.47,30.82,16.94,23.20
std,10.25,635.12,1.74,4282.04,16.91,13.64,12.08
min,5.00,4.00,1.00,0.00,7.00,2.00,6.00
25%,42.00,339.00,1.00,300.00,18.00,8.00,15.00
50%,51.00,595.00,1.00,700.00,25.00,13.00,20.00
75%,57.00,972.00,3.00,1900.00,38.00,21.00,30.00
max,76.00,3992.00,20.00,40425.00,105.00,105.00,118.00



--- sellers ---


,seller_zip_code_prefix
count,3095.00
mean,32291.06
std,32713.45
min,1001.00
25%,7093.50
50%,14940.00
75%,64552.50
max,99730.00



--- geolocation ---


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng
count,1000163.00,1000163.00,1000163.00
mean,36574.17,-21.18,-46.39
std,30549.34,5.72,4.27
min,1001.00,-36.61,-101.47
25%,11075.00,-23.60,-48.57
50%,26530.00,-22.92,-46.64
75%,63504.00,-19.98,-43.77
max,99990.00,45.07,121.11


In [7]:
print("=== NULL VALUE REPORT ===\n")
for name, df in tables.items():
    null_counts = df.isnull().sum()
    nulls = null_counts[null_counts > 0]
    print(f"--- {name} ---")
    if nulls.empty:
        print("  No null values found")
    else:
        for col, count in nulls.items():
            pct = (count / len(df)) * 100
            print(f"  {col}: {count:,} nulls ({pct:.1f}%)")
    print()

=== NULL VALUE REPORT ===

--- orders ---
  order_approved_at: 160 nulls (0.2%)
  order_delivered_carrier_date: 1,783 nulls (1.8%)
  order_delivered_customer_date: 2,965 nulls (3.0%)

--- order_items ---
  No null values found

--- customers ---
  No null values found

--- payments ---
  No null values found

--- reviews ---
  review_comment_title: 87,656 nulls (88.3%)
  review_comment_message: 58,247 nulls (58.7%)

--- products ---
  product_category_name: 610 nulls (1.9%)
  product_name_lenght: 610 nulls (1.9%)
  product_description_lenght: 610 nulls (1.9%)
  product_photos_qty: 610 nulls (1.9%)
  product_weight_g: 2 nulls (0.0%)
  product_length_cm: 2 nulls (0.0%)
  product_height_cm: 2 nulls (0.0%)
  product_width_cm: 2 nulls (0.0%)

--- sellers ---
  No null values found

--- geolocation ---
  No null values found

--- category_translation ---
  No null values found



In [8]:
print("=== DATA TYPES REPORT ===\n")
for name, df in tables.items():
    print(f"--- {name} ---")
    print(df.dtypes.to_string())
    print()

=== DATA TYPES REPORT ===

--- orders ---
order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object

--- order_items ---
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64

--- customers ---
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object

--- payments ---
order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64

--- reviews ---
review_id  

In [9]:
print("=== DUPLICATE ROWS REPORT ===\n")
for name, df in tables.items():
    dupes = df.duplicated().sum()
    print(f"  {name}: {dupes:,} duplicate rows")

=== DUPLICATE ROWS REPORT ===

  orders: 0 duplicate rows
  order_items: 0 duplicate rows
  customers: 0 duplicate rows
  payments: 0 duplicate rows
  reviews: 0 duplicate rows
  products: 0 duplicate rows
  sellers: 0 duplicate rows
  geolocation: 261,831 duplicate rows
  category_translation: 0 duplicate rows


In [10]:
print("=== UNIQUE VALUE CHECK ===\n")
print("orders.order_id unique:", orders['order_id'].nunique())
print("customers.customer_unique_id unique:", customers['customer_unique_id'].nunique())

=== UNIQUE VALUE CHECK ===

orders.order_id unique: 99441
customers.customer_unique_id unique: 96096


In [11]:
print("=== ORDER STATUS DISTRIBUTION ===\n")
print(orders['order_status'].value_counts())
print(f"\nTotal orders: {len(orders):,}")

=== ORDER STATUS DISTRIBUTION ===

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

Total orders: 99,441


In [12]:
print("=== REVIEW SCORE DISTRIBUTION ===\n")
print(reviews['review_score'].value_counts().sort_index())

=== REVIEW SCORE DISTRIBUTION ===

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64


In [13]:
print("=== PAYMENT TYPE DISTRIBUTION ===\n")
print(payments['payment_type'].value_counts())

print("\n=== INVALID INSTALLMENTS (value = 0) ===")
print(payments[payments['payment_installments'] == 0].shape[0], "rows with 0 installments")

=== PAYMENT TYPE DISTRIBUTION ===

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

=== INVALID INSTALLMENTS (value = 0) ===
2 rows with 0 installments


In [14]:
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'])

print("=== ORDER DATE RANGE ===")
print(f"  Earliest order: {orders['order_purchase_timestamp'].min()}")
print(f"  Latest order:   {orders['order_purchase_timestamp'].max()}")

=== ORDER DATE RANGE ===
  Earliest order: 2016-09-04 21:15:19
  Latest order:   2018-10-17 17:30:18


## Extraction Summary

All 9 raw tables loaded and inspected successfully.
No data was modified in this notebook.

Key issues identified for cleaning:
- Geolocation: 261,831 duplicate rows — deduplicate by zip code
- Orders: nulls in delivery timestamps — filter by order_status
- Payments: 2 invalid installment values + 3 not_defined payment types
- Products: 610 rows with missing category and dimensions
- Reviews: 88% null comment fields — drop comment columns
- Customers: 96,096 unique customers across 99,441 orders — 3,345 repeat buyers identified